Allow setting table AM for partition roots

huansong opened the pull request:
Ready to review.
----------

This PR includes two commits:

1. Using AM oid instead of AM handler oid in DefineRelation. Including it in this PR because: (c/p from my other review comment):
>It is not a must-have for this PR but it makes things easier and cleaner. W/o the change we'll have to check both pg_class and then pg_am to get the AM handler ID of the parent partition, despite that AM oid is all we need.
 
2. Support setting table AM for partition root:

Adapted from patch submitted against PG 15:
    https://postgr.es/m/CAE-ML%2B9NycfS3Q9PhMVpgB2NN62HMa9OBCwNW60AiO8NgCCNXw%40mail.gmail.com

This enables specification of a table AM for partition roots at table
    creation time. Creating a partitioned table with a specific AM specified
    will enable subsequent children to inherit the AM from the parent.

For setting the access method of an existing partitioned table:
* If the ONLY clause is specified, it will set the access method at
    the root only. Existing children will not be affected.
* If the ONLY clause is NOT specified, it will set the access method at
    the root and for all existing children.

Conflicts resolved:
* Mostly conflicts due to PG12 vs PG15: signature changes, non-existing
    macros etc.

* utility.c: Remove explicit sabotage of the am for partition roots that
    was in place.

* psql: Had to fix describe table for partitioned AO/AOCO tables:
    partitioned tables can't have AO/AOCO specific reloptions (yet)

Tests adjusted:
* alter_table_aocs2: partition root has AM in catalog.
* partition: partition root has AM in catalog.
* dsp: partition root has AM, and the child should default to the
	parent's AM instead of default_table_access_method.
* alter_distribution_policy: remove the statement which tests the case that
            you cannot changing AM for partition root, because now we CAN.
* create_am: specifying AM works now and child will inherit AM from root.

Dev pipeline:
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/atsetam_part

l-wang commented on src/test/regress/expected/alter_distribution_policy.out:
```diff
@@ -650,8 +650,6 @@ select * from atsdb order by 1, 2, 3;
  11 | 10 | 9
 (9 rows)
 
-alter table atsdb set with(appendonly = true);
```
@ashwinstar do we plan to at some point deprecate the Greenplum legacy syntax?

ashwinstar replied:
Yes, hoping deprecation notice of legacy syntax for some GPDB7 release year after its GA and removal for GPDB8. For sure not planning to continue further than GPDB8.

l-wang reviewed (commented):
I haven't finished reviewing the entire patch, but played with the following commands and it crashed:

```sql
create table foo (a int, b int) partition by range(b) (start (0) end (6) every (3));
alter table foo set access method ao_row; -- this sets the AM for the entire partitioning hierarchy as expected
create table foox (like foo);
alter table foo attach partition foox for values from (6) to (9); -- I'm not sure what should be the expectation for this one, but right now we keep the original AM of the attached partition as-is.
alter table foo set access method ao_row; -- Then I try to again set AM for the entire partitioning hierarchy, and this crashes
```

huansong commented:
> the following commands and it crashed:
create table foo (a int, b int) partition by range(b) (start (0) end (6) every (3));
alter table foo set access method ao_row; -- this sets the AM for the entire partitioning hierarchy as expected
create table foox (like foo);
alter table foo attach partition foox for values from (6) to (9); -- I'm not sure what should be the expectation for this one, but right now we keep the original AM of the attached partition as-is.
alter table foo set access method ao_row; -- Then I try to again set AM for the entire partitioning hierarchy, and this crashes

@l-wang Great catch! Fixed and add test case.

soumyadeep2007 commented on src/test/regress/expected/qp_orca_fallback_optimizer.out:
```diff
@@ -298,7 +298,7 @@ alter table ext_part attach partition p1 for values in (1);
 explain select * from ext_part;
 INFO:  GPORCA failed to produce a plan, falling back to planner
-DETAIL:  Feature not supported: Partitioned table with external/foreign partition
+DETAIL:  Feature not supported: External/foreign partition storage types
```
Couldn't quite grok this change.

huansong replied:
It's a different error message thrown by ORCA. Originally when trying to get the storage type of the partition root that has external child (`RetrieveRelStorageType()`), ORCA will throw `Feature not supported: Partitioned table with external/foreign partition` here: path https://github.com/greenplum-db/gpdb/blob/master/src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp#L2453 

Now because partition root carries a valid AM, we will return normally in `RetrieveRelStorageType()` but be caught in a similar check in the caller function `TranslateRTEToDXLLogicalGet()`, which throws `Feature not supported: External/foreign partition storage types`.

Originally I thought this is fine since it's essentially the same error. But now taking a relook I think we should better fix `RetrieveRelStorageType` because it's called in many other places too. We don't know if all caller have the same check as `TranslateRTEToDXLLogicalGet`. Therefore, now fixed it by handling partition table first in `RetrieveRelStorageType`. 

There's still one caveat though: the storage type returned in that function isn't necessarily the same as the AM of the partition root - see `RetrieveStorageTypeForPartitionedTable()` - ORCA checks the AMs of the child partition to get the storage type used by the caller. Should we use the AM of the root partition instead? - I'm not clear, but I think the best thing for now is to keep that logic consistent.

huansong commented on src/test/regress/expected/alter_table_set_am.out:
```diff
@@ -601,6 +601,61 @@ SELECT c.relname, a.amname, c.reloptions FROM pg_class c JOIN pg_am a ON c.relam
+-- Altering the storage model at the partition root should change the storage model for all existing children
+-- FIXME: currently children tables aren't inheriting reloptions of the root. After that is corrected, we 
+-- should add a correspondong test case for ALTER TABLE here too.
+CREATE TABLE atsetam_part(a int, b int) PARTITION BY RANGE(a);
```
Added.

l-wang reviewed (commented):
Could you explain more on why do we need the revert commit? I don't mind it but just curious what hurdle it introduced for this PR.

l-wang commented:
@huansong Could you also mention in the commit message about the expectations for `ALTER TABLE ... EXCHANGE PARTITION ...`? It looks like we keep the original storage type of the exchanged partitions, better to document it and add a test case for it.

huansong commented:
>Could you explain more on why do we need the revert commit? I don't mind it but just curious what hurdle it introduced for this PR.

It is not a must-have for this PR but it makes things easier and cleaner. W/o the change we'll have to check both pg_class and then pg_am to get the AM handler ID of the parent partition, despite that AM oid is all we need.

> Could you also mention in the commit message about the expectations for ALTER TABLE ... EXCHANGE PARTITION ...? It looks like we keep the original storage type of the exchanged partitions, better to document it and/or add a test case for it.

Good point. The EXCHANGE PARTITION behavior stays the same as existing, which is that the AMs of the source and target tables are exchanged. This PR only addresses issues with setting AM for **partition root**, so EXCHANGE PARTITION of the child table is unaffected. But yes we can clarify that in the commit message.

l-wang commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -2431,6 +2431,13 @@ CTranslatorRelcacheToDXL::RetrieveRelStorageType(Relation rel)
+	// is not necessarily the same as the one root partition carries
+	if (rel->rd_rel->relkind == RELKIND_PARTITIONED_TABLE)
+	{
+		return RetrieveStorageTypeForPartitionedTable(rel);
```
OK I'll keep an note on this we can revisit after this PR is merged.

l-wang commented:
> > Could you also mention in the commit message about the expectations for ALTER TABLE ... EXCHANGE PARTITION ...? It looks like we keep the original storage type of the exchanged partitions, better to document it and/or add a test case for it.
> 
> Good point. The EXCHANGE PARTITION behavior stays the same as existing, which is that the AMs of the source and target tables are exchanged. This PR only addresses issues with setting AM for **partition root**, so EXCHANGE PARTITION of the child table is unaffected. But yes we can clarify that in the commit message.

Sorry my bad! Right this doesn't have much to do with `EXCHANGE PARTITION`, that was a typo I meant to say `ALTER TABLE ... ATTACH PARTITION ...`

huansong commented:
> Sorry my bad! Right this doesn't have much to do with EXCHANGE PARTITION, that was a typo I meant to say ALTER TABLE ... ATTACH PARTITION ...

No worries! I see. Yes ATTACH PARTITION is more relevant to this PR. The current behavior is that the attached partition will keep its original AM. I believe this is the desired behavior because the change of AM shouldn't happen outside of ALTER TABLE (i.e. altering the leaf table being attached). Imagine EXCHANGE PARTITION but just w/o a leaf partition to swap, the new leaf should keep its original AM.

Agreed that we should add commit message and test case. Done.

huansong commented:
CI failure is an unrelated flake. Dev pipeline all passed: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/atsetam_part

l-wang commented:
> > Could you explain more on why do we need the revert commit? I don't mind it but just curious what hurdle it introduced for this PR.
> 
> It is not a must-have for this PR but it makes things easier and cleaner. W/o the change we'll have to check both pg_class and then pg_am to get the AM handler ID of the parent partition, despite that AM oid is all we need.

Thanks for explaining. After discussing w/ @soumyadeep2007 and @huansong offline, I'm convinced reverting that commit makes our lives easier. Given upstream doesn't check table am handler id for heap am handler either, the previous commit was likely added for our own testing purposes only at the time of development.
